### ISSUE TITLE
Monit doesn't restart an unresponsive Cassandra process

### ISSUE THREAD

**Issue by rkd-msw** [Metaswitch/clearwater-cassandra#118]
#### Symptoms

I'd expect Monit to restart Cassandra if Cassandra was running but not listening on port 9160. However, after the testing described in https://github.com/Metaswitch/clearwater-etcd/issues/375, I noticed that:

* monit was reporting poll_cassandra as "Status failed"
* cassandra wasn't listening on port 9160
* the cassandra process was last restarted over an hour ago

```
clearwater@rkd-ipv6-homestead-1:~$ sudo monit summary | grep cassandra
 cassandra_process                Running                     Process
 cassandra_uptime                 Status ok                   Program
 poll_cassandra                   Status failed               Program
```

```
clearwater@rkd-ipv6-homestead-1:~$ sudo /usr/share/clearwater/bin/run-in-signaling-namespace netstat -tnlp
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 0.0.0.0:5658            0.0.0.0:*               LISTEN      26532/homestead
tcp        0      0 0.0.0.0:3868            0.0.0.0:*               LISTEN      26532/homestead
tcp6       0      0 127.0.0.1:40906         :::*                    LISTEN      5386/java
tcp6       0      0 ::1:8888                :::*                    LISTEN      26532/homestead
tcp6       0      0 fd5f:5d21:845:1422:8888 :::*                    LISTEN      26532/homestead
tcp6       0      0 :::5658                 :::*                    LISTEN      26532/homestead
tcp6       0      0 :::3868                 :::*                    LISTEN      26532/homestead
tcp6       0      0 127.0.0.1:7199          :::*                    LISTEN      5386/java
```

```
clearwater@rkd-ipv6-homestead-1:~$ date
Tue Dec 13 12:52:18 UTC 2016
clearwater@rkd-ipv6-homestead-1:~$ ps aux | grep cassa
root      4539  0.1  0.5 471288 21368 ?        Sl   11:37   0:04 /usr/share/clearwater/clearwater-cluster-manager/env/bin/python /usr/share/clearwater/bin/clearwater-cluster-manager --mgmt-local-ip=10.225.166.92 --sig-local-ip=fd5f:5d21:845:1422:1::100 --local-site=site1 --remote-site= --remote-cassandra-seeds= --signaling-namespace=signaling --uuid=6d9ebbfd-b570-4eac-b956-c7c030656337 --etcd-key=clearwater --etcd-cluster-key=homestead --cluster-manager-enabled=Y --log-level=2 --log-directory=/var/log/clearwater-cluster-manager --pidfile=/var/run/clearwater-cluster-manager.pid
cassand+  5386  0.6 35.5 1761796 1438636 ?     SLl  11:38   0:28 /usr/lib/jvm/java-7-openjdk-amd64//bin/java -ea -javaagent:/usr/share/cassandra/lib/jamm-0.3.0.jar -XX:+CMSClassUnloadingEnabled -XX:+UseThreadPriorities -XX:ThreadPriorityPolicy=42 -Xms1024M -Xmx1024M -Xmn200M -XX:+HeapDumpOnOutOfMemoryError -Xss256k -XX:StringTableSize=1000003 -XX:+UseParNewGC -XX:+UseConcMarkSweepGC -XX:+CMSParallelRemarkEnabled -XX:SurvivorRatio=8 -XX:MaxTenuringThreshold=1 -XX:CMSInitiatingOccupancyFraction=75 -XX:+UseCMSInitiatingOccupancyOnly -XX:+UseTLAB -XX:CompileCommandFile=/etc/cassandra/hotspot_compiler -XX:CMSWaitDuration=10000 -XX:+CMSParallelInitialMarkEnabled -XX:+CMSEdenChunksRecordAlways -XX:CMSWaitDuration=10000 -XX:+UseCondCardMark -Dcassandra.jmx.local.port=7199 -XX:+DisableExplicitGC -Dcassandra.ignore_rack=true -Dcassandra.ignore_dc=true -Dlogback.configurationFile=logback.xml -Dcassandra.logdir=/var/log/cassandra -Dcassandra.storagedir= -Dcassandra-pidfile=/var/run/cassandra/cassandra.pid -cp /etc/cassandra:/usr/share/cassandra/lib/ST4-4.0.8.jar:/usr/share/cassandra/lib/airline-0.6.jar:/usr/share/cassandra/lib/antlr-runtime-3.5.2.jar:/usr/share/cassandra/lib/commons-cli-1.1.jar:/usr/share/cassandra/lib/commons-codec-1.2.jar:/usr/share/cassandra/lib/commons-lang3-3.1.jar:/usr/share/cassandra/lib/commons-math3-3.2.jar:/usr/share/cassandra/lib/compress-lzf-0.8.4.jar:/usr/share/cassandra/lib/concurrentlinkedhashmap-lru-1.4.jar:/usr/share/cassandra/lib/disruptor-3.0.1.jar:/usr/share/cassandra/lib/guava-16.0.jar:/usr/share/cassandra/lib/high-scale-lib-1.0.6.jar:/usr/share/cassandra/lib/jackson-core-asl-1.9.2.jar:/usr/share/cassandra/lib/jackson-mapper-asl-1.9.2.jar:/usr/share/cassandra/lib/jamm-0.3.0.jar:/usr/share/cassandra/lib/javax.inject.jar:/usr/share/cassandra/lib/jbcrypt-0.3m.jar:/usr/share/cassandra/lib/jline-1.0.jar:/usr/share/cass ndra/lib/jna-4.0.0.jar:/usr/share/cassandra/lib/json-simple-1.1.jar:/usr/share/cassandra/lib/libthrift-0.9.2.jar:/usr/share/cassandra/lib/logback-classic-1.1.2.jar:/usr/share/cassandra/lib/logback-core-1.1.2.jar:/usr/share/cassandra/lib/lz4-1.2.0.jar:/usr/share/cassandra/lib/metrics-core-2.2.0.jar:/usr/share/cassandra/lib/netty-all-4.0.23.Final.jar:/usr/share/cassandra/lib/reporter-config-2.1.0.jar:/usr/share/cassandra/lib/slf4j-api-1.7.2.jar:/usr/share/cassandra/lib/snakeyaml-1.11.jar:/usr/share/cassandra/lib/snappy-java-1.0.5.2.jar:/usr/share/cassandra/lib/stream-2.5.2.jar:/usr/share/cassandra/lib/super-csv-2.1.0.jar:/usr/share/cassandra/lib/thrift-server-0.3.7.jar:/usr/share/cassandra/apache-cassandra-2.1.15.jar:/usr/share/cassandra/apache-cassandra-thrift-2.1.15.jar:/usr/share/cassandra/apache-cassandra.jar:/usr/share/cassandra/cassandra-driver-core-2.0.9.2.jar:/usr/share/cassandra/netty-3.9.0.Final.jar:/usr/share/cassandra/stress.jar: -XX:HeapDumpPath=/var/lib/cassandra/java_1481629088.hprof -XX:ErrorFile=/var/lib/cassandra/hs_err_1481629088.log org.apache.cassandra.service.CassandraDaemon
```

#### Impact

Cassandra problems may cause an outage that requires manual intervention to recover.

#### Release and environment 

`10.0.0-161212-~255.00.0-161212.145405`, using IPv6 and only one Homestead node.

#### Steps to reproduce

Run `mark_node_failed` to shut down Cassandra on a node, then see it not restart.

**Comment by eleanor-merry** [comment]
Looks like https://issues.apache.org/jira/browse/CASSANDRA-9223 (which is an example of a Cassandra bug that's only fixed in 3+)

**Comment by eleanor-merry** [comment]
We could work around this by having poll_cassandra more aggressively kill cassandra, rather than relying on cassandra stop.

**Comment by eleanor-merry** [comment]
This should be fixed by the linked PR; passing back for retest

**Comment by rkd-msw** [comment]
Yep, fixed on `10.0.0-170202-~255.00.0-170202.020438`. I tested with both the original `mark_node_failed` stuff and also `kill -STOP` (where the cqlsh stuff kicked in and killed it for unresponsiveness).
